lib/src/client/transport/xhr_transport.dart

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2018, the gRPC project authors. Please see the AUTHORS file
+// for details. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+import 'dart:html';
+import 'dart:typed_data';
+
+import 'package:meta/meta.dart';
+
+import '../../shared/message.dart';
+import 'transport.dart';
+import 'web_streams.dart';
+
+class XhrTransportStream implements GrpcTransportStream {
+  HttpRequest _request;
+  int _requestBytesRead = 0;
+  StreamController<ByteBuffer> _incomingProcessor;
+  StreamController<GrpcMessage> _incomingMessages;
+  StreamController<List<int>> _outgoingMessages;
+
+  @override
+  Stream<GrpcMessage> get incomingMessages => _incomingMessages.stream;
+
+  @override
+  StreamSink<List<int>> get outgoingMessages => _outgoingMessages.sink;
+
+  XhrTransportStream(this._request) {
+    _incomingProcessor = StreamController();
+    _incomingMessages = StreamController();
+    _outgoingMessages = StreamController();
+
+    _incomingProcessor.stream
+        .transform(GrpcWebDecoder())
+        .transform(grpcDecompressor())
+        .listen(_incomingMessages.add,
+            onError: _incomingMessages.addError,
+            onDone: _incomingMessages.close);
+
+    _outgoingMessages.stream
+        .map(frame)
+        .listen((data) => _request.send(data));
+
+    _request.onReadyStateChange.listen((data) {
+      final contentType = _request.getResponseHeader('Content-Type');
+      if (contentType == null) {
+        return;
+      }
+
+      if (_request.readyState == HttpRequest.HEADERS_RECEIVED) {
+        if (contentType.startsWith('application/grpc')) {
+          if (_request.response == null) {
+            return;
+          }
+
+          // Force a metadata message with headers
+          final headers = GrpcMetadata(_request.responseHeaders);
+          _incomingMessages.add(headers);
+        }
+      }
+
+      if (_request.readyState == HttpRequest.DONE) {
+        _incomingProcessor.close();
```

Comment:
What if `_request.onError` happened? Perhaps the connection was broken... If this was a streaming response would we know the difference between a broken connection and a connection that closed because no more messages was supposed to arrive?